mmrotate/models/utils/align.py

```diff
@@ -0,0 +1,197 @@
+        dw, dh = w / self.kernel_size, h / self.kernel_size
+        x, y = dw[:, None] * xx, dh[:, None] * yy
+        xr = cos[:, None] * x - sin[:, None] * y
+        yr = sin[:, None] * x + cos[:, None] * y
+        x_anchor, y_anchor = xr + x_ctr[:, None], yr + y_ctr[:, None]
+        # get offset filed
+        offset_x = x_anchor - x_conv
+        offset_y = y_anchor - y_conv
+        # x, y in anchors is opposite in image coordinates,
+        # so we stack them with y, x other than x, y
+        offset = torch.stack([offset_y, offset_x], dim=-1)
+        offset = offset.reshape(anchors.size(0),
+                                -1).permute(1, 0).reshape(-1, feat_h, feat_w)
+        return offset
+
+    def forward_single(self, x, anchor, stride):
+        """Forward function for single level."""
+        num_imgs, _, H, W = x.shape
+        offset_list = [
+            self.get_offset(anchor[i].reshape(-1, 5), (H, W), stride)
+            for i in range(num_imgs)
+        ]
+        offset_tensor = torch.stack(offset_list, dim=0)
+        x = self.deform_conv(x, offset_tensor)
+        x = self.relu(x)
+        return x
+
+    def forward(self, x: List[Tensor],
+                anchors: List[List[Tensor]]) -> List[Tensor]:
+        """Forward function."""
+        mlvl_anchors = []
+        for i in range(len(x)):
+            anchor = torch.stack([anchor[i] for anchor in anchors], dim=0)
+            mlvl_anchors.append(anchor)
+        out = []
+        for x, anchor, stride in zip(x, mlvl_anchors, self.strides):
+            out.append(self.forward_single(x, anchor, stride))
+        return out
+
+
+@TASK_UTILS.register_module()
+class PseudoAlignModule(BaseModule):
+    """Pseudo Align Module."""
+
+    def forward(self, x: List[Tensor],
+                anchors: List[List[Tensor]]) -> List[Tensor]:
+        """Forward function."""
+        return x
+
+
+@TASK_UTILS.register_module()
+class DCNAlignModule(DeformConv2dPack):
+    """DCN Align Module.
+
+    All args are from DeformConv2dPack.
+    TODO: maybe use build_conv_layer is more flexible.
+    """
+
+    def forward(self, x: List[Tensor],
+                anchors: List[List[Tensor]]) -> List[Tensor]:
+        """Forward function."""
+        return super(DCNAlignModule, self).forward(x)
+
+
+@TASK_UTILS.register_module()
+class FRM(BaseModule):
+    """Feature refine module for `R3Det`.
+
+    Args:
+        feat_channels (int): Number of input channels.
+        strides (list[int]): The strides of featmap.
+        conv_cfg (:obj:`ConfigDict` or dict, optional): Config dict for
+            convolution layer. Defaults to None.
+        norm_cfg (:obj:`ConfigDict` or dict, optional): Config dict for
+            normalization layer. Defaults to None.
+    """
+
+    def __init__(self,
+                 feat_channels: int,
+                 strides: List[int],
+                 conv_cfg: OptConfigType = None,
+                 norm_cfg: OptConfigType = None) -> None:
```

Comment:
is 'conv_cfg' and 'norm_cfg' useful here? maybe it canbe remove or use `ConvModule` in `init_layers`?